File: JavaMinMq/src/main/java/com/ajrzeznik/AddressMap.java
package com.ajrzeznik;

import java.util.HashMap;
import java.util.Map;

public class AddressMap {

    private final String nodeName;
    // TODO AR: Thread safety is probably NOT needed here, as each publicher can hold a pubsocket as a reference, and
    // the pubsocket itself is guaranteed to be threadsafe. But this assumption should be checked.
    final HashMap<String, PubSocket> socketMap = new HashMap<>();
    private boolean allConnected = false;

    AddressMap(String nodeName) {
        this.nodeName = nodeName;
    }

    // Returns true if a new address was added.
    boolean updateAddress(String name, String address) {
        if (socketMap.containsKey(name)) {
            PubSocket socket = socketMap.get(name);
            if (socket.getAddress().equals(address)){
                return false;
            }
            socket.updateAddress(address);

        } else {
            socketMap.put(name, PubSocket.create(address));
        }
        allConnected = false;
        return true;
    }

    boolean allNewlyConnected() {
        if (allConnected) {
            return false;
        } else {
            boolean connectedCheck = true;
            //TODO AR: Optimize with a break out early?
            for (Map.Entry<String,PubSocket> item : socketMap.entrySet()) {
                connectedCheck = connectedCheck && item.getValue().isConnected();
            }
            allConnected = connectedCheck;
            return connectedCheck;
        }
    }
}

File: min_mq/src/address_map.rs
use std::borrow::Borrow;
use std::cell::RefCell;
use std::collections::HashMap;
use std::rc::Rc;
use crate::sockets::PubSocket;

pub(crate) struct AddressMap {
    all_connected: bool,
    node_name: String,
    pub(crate) socket_map: Rc<RefCell<HashMap<String, PubSocket>>>
}

impl AddressMap {
    pub fn new(name: &str) -> Self{
        let mut socket_map = HashMap::new();
        socket_map.insert(name.to_string(), PubSocket::new("tcp://localhost:55555"));

        AddressMap {
            all_connected: false,
            node_name: name.to_string(),
            socket_map: Rc::new(RefCell::new(socket_map))
        }
    }
    //TODO AR: Refcell stuff here is pretty ugly, probably should be fixed
    pub fn update_address(&mut self, name: &str, address: &str) -> bool{
        let mut current_map = self.socket_map.borrow_mut();
        if current_map.contains_key(name){
            let socket = current_map.get_mut(name).unwrap();
            if socket.get_address() == address {
                return false;
            }
            socket.update_address(address);
        } else {
            current_map.insert(name.to_string(), PubSocket::new(address));
        }
        true
    }

    pub fn all_newly_connected(&mut self) -> bool {
        if self.all_connected {
            false
        } else {
            let ref_current_map: &RefCell<HashMap<String, PubSocket>> = self.socket_map.borrow();
            let current_map = ref_current_map.borrow();
            self.all_connected = !current_map.iter().any(|item| !item.1.get_connected());
            self.all_connected
        }
    }
}